runtime/onert/backend/train/ops/DepthwiseConvolutionLayer.cc

```diff
@@ -0,0 +1,194 @@
+    _use_padded_filter{false}, _padded_filter{nullptr}, _filter_buffers{nullptr},
+    _filter_dim_buffers{nullptr}, _dconv_kernel{new nnfw::cker::train::DepthwiseConv()}
+{
+  // DO NOTHING
+}
+
+void DepthwiseConvolutionLayer::configure(
+  const IPortableTensor *input, const IPortableTensor *kernel, const IPortableTensor *bias,
+  IPortableTensor *output, IPortableTensor *back_prop_input, IPortableTensor *grad_weights,
+  IPortableTensor *grad_bias, const IPortableTensor *back_prop_output, const uint32_t paddingLeft,
+  const uint32_t paddingRight, const uint32_t paddingTop, const uint32_t paddingBottom,
+  const uint32_t strideWidth, const uint32_t strideHeight, const uint32_t multiplier,
+  const uint32_t dilationWidth, const uint32_t dilationHeight, const ir::Activation activation,
+  const std::shared_ptr<ExternalContext> &external_context)
+{
+  cpu::ops::DepthwiseConvolutionLayer::configure(
+    input, kernel, bias, paddingLeft, paddingRight, paddingTop, paddingBottom, strideWidth,
+    strideHeight, multiplier, dilationWidth, dilationHeight, activation, output, external_context);
+  _back_prop_input = back_prop_input;
+  _back_prop_output = back_prop_output;
+  _grad_weights = grad_weights;
+  _grad_bias = grad_bias;
+
+  if (activation != ir::Activation::NONE)
+  {
+    _act_back_prop_output =
+      std::make_unique<BackPropTensor>(_back_prop_output->get_info(), _back_prop_output->layout());
+    _act_back_prop_output->setBuffer(
+      std::make_shared<basic::Allocator>(_act_back_prop_output->total_size()));
+  }
+
+  const int64_t k_packet_size = [&]() {
+    const auto data_type = _back_prop_output->data_type();
+    switch (data_type)
+    {
+      case OperandType::FLOAT32:
+      {
+        return _dconv_kernel->kPacketSize<float>();
+      }
+      default:
+        throw std::runtime_error("train DepthwiseConvolutionLayer: unsupported data type");
+    }
+  }();
+
+  const auto incoming_shape = getShape(_back_prop_output);
+  const auto filter_shape = getShape(_kernel);
+  const int batch = incoming_shape.Dims(0);
+  const int out_depth = incoming_shape.Dims(3);
+  const int filter_rows = filter_shape.Dims(1);
+  const int filter_cols = filter_shape.Dims(2);
+
+  const int filter_spatial_size = filter_rows * filter_cols;
+  const int padded_filter_inner_dim_size =
+    ((out_depth + k_packet_size - 1) / k_packet_size) * k_packet_size;
+
+  _use_padded_filter = (out_depth % k_packet_size) == 0 ? false : true;
+
+  // prepare padded_filter buffer for cker
+  auto padded_filter_info = ir::OperandInfo(_kernel->get_info());
+  padded_filter_info.shape({batch, filter_spatial_size, padded_filter_inner_dim_size});
+  _padded_filter = std::make_unique<Tensor>(padded_filter_info, _kernel->layout());
+  _padded_filter->setBuffer(std::make_shared<basic::Allocator>(_padded_filter->total_size()));
+
+  // prepare out_bprop and in_bprop buffer for cker
+  const int thread_count = _dconv_kernel->getThreadCount();
```

Comment:
(just for curiosity)

Does this `thread_count` mean that - how many threads are utilized in ckernel?

Reply:
Aha, I read the `DepthwiseConv` kernel code.

Reply:
@zetwhite

Yes, you're right. The depthwise conv kernel allocates additional memory for each thread. This code pre-allocates the kernel memory in advance instead of allocating memory every time in each thread. And allocate memory as much as the number of threads to avoid collisions that occurs simultaneously in multiple threads.